ide/project.dependency/src/org/netbeans/modules/project/dependency/reload/Reloader.java

```diff
@@ -383,163 +385,259 @@ private void markInconsistentParts() {
+     * It protects against repeated loads: if the same files are modified, or the same implementations request reload as the last time, the
+     * operation fails with OUT_OF_SYNC (file inconsistencies) or 
+     * @param variant
+     * @return 
+     */
+    private CompletableFuture<ProjectState> maybeMakeRetry(Collection variant) {
+        // special case: if some of the phases reported ERROR = code error, bail out and do not even try to retry
+        if (loadData.stream().anyMatch(ctx -> ctx.reportLoadError)) {
+            return null;
+        }
         Collection<LoadContextImpl> retries;
         boolean forceReload = loadData.stream().anyMatch(d -> d.reloadRequested);
-        markInconsistentParts();
         retries = loadData.stream().filter(d -> d.reloadRequested).collect(Collectors.toList());
-        Throwable error = null;
+        Collection<LoadContextImpl> fileInconsistencies = new HashSet<>();
+        checkFileTimestamps(fileInconsistencies);
+        boolean inconsistentRetry = request.isConsistent() && !fileInconsistencies.isEmpty();
+        if (inconsistentRetry) {
+            if (!lastInconsistent.isEmpty() && fileInconsistencies.containsAll(lastInconsistent)) {
+                ProjectOperationException ex = new ProjectOperationException(project, ProjectOperationException.State.OUT_OF_SYNC, Bundle.ERR_ProjectModifiedWhileLoading(ProjectUtils.getInformation(project).getDisplayName()));
+                registry.createState(originalState, project, variant, parts, false, request);
+                return CompletableFuture.failedFuture(ex);
+            }
+        } else if (retries.isEmpty()) {
+            return null;
+        } else if (!lastRetries.isEmpty() && retries.containsAll(lastRetries)) {
+            // too bad: we seem to be in a retry cycle. Bail out.
+            LOG.log(Level.WARNING, "Project {0} is reloading repetadely. The following provider(s) reload in a loop: {1}", new Object[]{project.getProjectDirectory(), lastRetries});
+            // FIXME: create a ProjectState from the data available so far.
+            ProjectOperationException ex = new ProjectOperationException(project, ProjectOperationException.State.ERROR, Bundle.ERR_ReloadingLoop(ProjectUtils.getInformation(project).getDisplayName(), lastRetries));
+            registry.createState(originalState, project, variant, parts, false, request);
+            return CompletableFuture.failedFuture(ex);
+        } 
         if (LOG.isLoggable(Level.FINE)) {
-            LOG.log(Level.FINE, "{0} load round completed.", toString());
+            String s = Stream.concat(fileInconsistencies.stream(),retries.stream()).
+                    map(c -> c.impl.getClass().getName()).collect(Collectors.joining(", "));
+            if (LOG.isLoggable(Level.FINE)) {
+                LOG.log(Level.FINE, "{0} reloads again because of {1}", new Object[]{toString(), s});
+            }
         }
```

Comment:
if i see this correctly on gh, the inner guard is redundant

Reply:
addressed in 2d925f2